### PR TITLE
Altered if clasuses around selection of commands in umi_tools.py

### PR DESCRIPTION
**Pull request by IanSudbery**
Fixes (or at least starts to fix) #536 

@TomSmithCGAT Bit confused as to what is going on in the umi_tools module. Specifically what the original purpose of the if clause at lines 51-52 was.

I've also made each tool print its usage message if no options are passed at all. Don't know if this was the right was to do this, or if I should have done something more central (although that would become a problem if there was a tool that could be run without options. Which currently there isn't).

### CODE REVIEW

File: umi_tools/Utilities.py [hunk 1]
```diff
@@ -1030,7 +1030,7 @@ def Start(parser=None,
 
     if return_parser:
         return parser
-
+
```
Comment (TomSmithCGAT) [inline umi_tools/Utilities.py:1033]: Is that just a tab added?

File: umi_tools/count_tab.py [hunk 2]
```diff
@@ -86,12 +86,17 @@ def main(argv=None):
 
     if argv is None:
         argv = sys.argv
-
+
```
Comment (TomSmithCGAT) [inline umi_tools/count_tab.py:89]: Another tab to remove?

File: umi_tools/count_tab.py [hunk 3]
```diff
@@ -86,12 +86,17 @@ def main(argv=None):
...
     if argv is None:
         argv = sys.argv
-
+        
     # setup command line parser
     parser = U.OptionParser(version="%prog version: $Id$",
                             usage=usage,
                             description=globals()["__doc__"])
 
+    if len(argv) == 1:
```
Comment (TomSmithCGAT) [inline umi_tools/count_tab.py:95]: Happy with this to catch use of tool with no options

File: umi_tools/umi_tools.py [hunk 4]
```diff
@@ -50,12 +48,22 @@ def main(argv=None):
...
         print("UMI-Tools: Version %s" % __version__)
 
     command = argv[1]
 
-    module = importlib.import_module("umi_tools." + command, "umi_tools")
+    try:
+        module = importlib.import_module("umi_tools." + command, "umi_tools")
+    except ImportError:
+        print("Command %s not recognosied" % command)
+        print()
```
Comment (TomSmithCGAT) [inline umi_tools/umi_tools.py:60]: ?

File: umi_tools/umi_tools.py [hunk 5]
```diff
@@ -50,12 +48,22 @@ def main(argv=None):
 
         return
 
-    elif argv[2] in ["--help", "-h", "--help-extended"]:
+    elif len(argv) > 2 and  argv[2] in ["--help", "-h", "--help-extended"]:
```
Comment (TomSmithCGAT) [inline umi_tools/umi_tools.py:51]: I think this elif can be removed. The tools all have -h/--help/--help-extended options so no need to catch anything here

**Comment by TomSmithCGAT** [comment]
Hey @IanSudbery. Assuming you're referring to and the line numbers are from this PR?
https://github.com/CGATOxford/UMI-tools/blob/0c72fa99c518088aa4505ee16721f85ac42eb0c2/umi_tools/umi_tools.py#L53-L54

I'm not sure what was going on with that elif either. It looks like I added it when we split into multiple tools to print help when a tool is specified and then you changed the output to be the version number (0101494bf236a485cbb14a55255113374df3c9a9 ). Regardless, each tool command line has help options added so no need to catch it in `umi_tools.py`.

I'm fine with having the same code in each tool to output help if no options are given. I agree it would be best done centrally though if you can be bothered recoding. I guess the easiest way to do it would be to catch instances where `len(sys.args)==2` after the checks for version/help options, print the line you have about needing to add required options and then add `--help` to the sys.args and call the tool module main. I can't imagine us adding a tool with no help in the future since our base arg parser adds `--help` anyway

**Comment by TomSmithCGAT** [comment]
Can't see how to comment on lines that haven't been updated in the PR but L46 in `umi_tools.py`:

`elif len(argv) == 1 or argv[1] == "--version" or argv[1] == "-v":`

That `len(argv) == 1` can go, since that case is met by the `if` and causes the help to be printed.

Minor thing.

File: umi_tools/count_tab.py (hunk 2 above)
Comment (TomSmithCGAT) [reply umi_tools/count_tab.py:89]: There are a few of these in the PR. One in each tool it looks like. Unless I'm missunderstanding the github diff markup.

File: umi_tools/umi_tools.py (hunk 4 above)
Comment (TomSmithCGAT) [reply umi_tools/umi_tools.py:60]: What's the `print()` for?

File: umi_tools/count_tab.py (hunk 2 above)
Comment (IanSudbery) [reply umi_tools/count_tab.py:89]: I don't think you are misunderstanding.

File: umi_tools/umi_tools.py (hunk 4 above)
Comment (IanSudbery) [reply umi_tools/umi_tools.py:60]: Lazy way to add a blank line 😬

File: umi_tools/umi_tools.py (hunk 5 above)
Comment (IanSudbery) [reply umi_tools/umi_tools.py:51]: My guess is that this was added because the standard `--help` within each tool doesn't output the version number?

**Comment by IanSudbery** [comment]
>  can't imagine us adding a tool with no help in the future since our base arg parser adds --help anyway

Thats not what I meant. I guess I was thinking about a tool that had no required arguements. For example, you could imagine that if `extract` had a default `--pattern` it could be run:

```
cat my_reads.fastq | umi_tools extract > my_reads.extracted.fastq
```

Although of course it doesn't, and I think all our tools require options of some sort (although come to think of it the new `prepare_for_rsem` or whatever we are going to call it might not). 

In the above `len(argv) == 1`, but there is nothing wrong with it. Although perhaps this is just premature optimisation on my part. 

I'll clean up the rough tabs.

File: umi_tools/umi_tools.py (hunk 5 above)
Comment (TomSmithCGAT) [reply umi_tools/umi_tools.py:51]: Ah, maybe. 

On a related note, the `--version` options for each tool also don't properly return the version. Should probably just remove that option from the tools and make it available via `umi_tools --version` only

**Comment by IanSudbery** [comment]
Whitespace should be tidied up.

**Review by TomSmithCGAT — APPROVED** [review]

**Comment by TomSmithCGAT** [comment]
Noticed in passing that a lot of the changes to umi_tools.py are also in #506. Should we first merge that one so that the contributions include @[epruesse]?

**Comment by IanSudbery** [comment]
yes
Ian Sudbery
Senior Lecturer in Bioinformatics,
Sheffield Institute for Nucleic Acids,
School of Biosciences,
The University of Sheffield.

web: www.sudlab.co.uk
Tel: 0114 222 2738
Twitter: IanSudbery
Show Calendar Availability
<https://calendar.google.com/calendar/u/0?cid=aS5zdWRiZXJ5QHNoZWZmaWVsZC5hYy51aw>

**Comment by TomSmithCGAT** [comment]
@IanSudbery - This is OK to merge, right?

**Pull request merged**